### ISSUE TITLE
Low CPU use (28%)

### ISSUE THREAD

**Issue by Alasir** [Microsoft/BashOnWindows#1626]
I could not find this anywhere, but if there is a solution/workaround, please inform me.

I'm running OpenFoam (A numerical simulation program) using Bash, and is less than impressed with the computing time. Task Manager tells me only 28% (varies +/- 3%) of the CPU is being used, and I would like to increase this as much as possible. Infact, I would like this simulation to take priority over most other non-essential processes. Total CPU workload is displayed as around 40% while the program is running.

Win10Home v1607
Build: 14393.693

CPU: i5core 2.7GHz
RAM: 8GB DDR3

Steps:
To do this spesific simulation:
-Install Bash as per Microsoft instructions.
-Install OpenFoam v4.1 as per instructions from www.openfoam.org
-run the pisoFoam pitzDaily tutorial simulation (need root). To do so, go to the folder for pitzDaily and run "blockMesh" then "pisoFoam"

**Comment by aseering** [comment]
Hi @Alasir -- just checking, how many cores does your processor have?  28% CPU kind of sounds like "fully utilizing one core" to me, on some hardware.

**Comment by Alasir** [comment]
Processor: Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz, 2701 Mhz, 2 Core(s), 4 Logical Processor(s)

**Comment by aseering** [comment]
Hi again -- I thought I had replied but I don't see the reply here; sorry if this is a dupe.

If you have 4 logical cores and are using just over 25% of your CPU, that sounds like you're just using one core.  

I don't know much about OpenFoam, but a quick search implies that it will only use one core by default.  It has to be configured to partition your workload, for example as follows:

http://web.student.chalmers.se/groups/ofw5/Basic_Training/runningInParallelLucchini.pdf

**Comment by twocs** [comment]
Similarly, I don't know about OpenFoam, but the basic step I would take would be to check if Ubuntu will run that demo on multiple cores. The question is really better asked of OpenFoam community, but I'll take a shot at it. Note that BashOnWindows does not support GPU computing. I'd say that's a big hint that it's not ready for serious parallel computing work.

That being said, let's say you are eager to run a demo on multiple cores in WSL. Another step would be find to a demo that specifically states that it's run in parallel. This is the first one I found:  https://bitbucket.org/mattshax/parallel_openfoam

I also see that a lot of different implementations of OpenFoam. My last suggestion would be to use one that explicitly states it is compiled with multi-core and Windows support, for example: http://joomla.bluecape.com.pt/index.php?option=com_content&task=view&id=18&Itemid=30

**Comment by crcrewso** [comment]
To add to this. I've done some other Monte Carlo on WSL work and have found that Bash can see multiple cores, though I had to use 'at' as my scheduler so it's not something I do regularly. I second the look into parallel openfoam. As of this point I have not tried any openMP or MPI multithreading but my suspicion is that those would work as expected too. 

This was on production system using 10586, not an insider build.

**Comment by benhillis** [comment]
This discussion has run its course.

**Comment by jnmlujnmlu** [comment]
Hi, 

there are a coupe of points, that I figured out, but there is still a lot of work to do with OpenFOAM in WSL. 

First, _ALWAYS_ write the output of OpenFOAM into a log file. The output into the bash makes the simulation considerably slower. 

Second, I found, that the Windows defender is not very fond of the fact, that OpenFOAM writes out all those files onto the disk and always occupies one CPU, when OpenFOAM writes. This is bad, if your save a lot of time steps. No idea, why this happens. I usually don't write out the time-steps (or only a few of them) and use functionObjects to have a small I/O. 

Third, some of the OpenFOAM solvers seem to have a better parallel performance than others even in Ubuntu. The best idea is (if possible) to go and test the scaling on a native installation of a linux server edition, run the scaling test there and compare the scaling to WSL.

I hope this helps. I'd love to hear the opinion of others about parallel scaling of OpenFOAM in WSL.

József
